### PR TITLE
feat(db): database migrations are now managed by phinx

### LINKED ISSUE #5442: Allow upgrade.php to perform schema upgrades before booting core
Currently upgrade.php begins by booting the core, which means that the system has to handle old schemas, in theory back to 1.0. This makes the core much more complex than it needs to be. E.g., we must be careful not to reference certain columns in queries.

I suggest an early process in upgrade.php designed just to handle schema upgrades before boot, so that the core can be further optimized for a single schema.

This may be needed if we ever move on #2675.

### LINKED ISSUE #7947: Make schema migrations simpler
Data migrations are probably never going to be easy, but I think we can make schema updates trivial. Closest tool I've seen so far is schematic (https://github.com/andrefigueira/Schematic), but I don't like some of the API decisions and it is licensed gpl3, not compatible with MIT core.

I do like the idea of defining schema in json and having the tool figure out what diffs to make, though.

I started a gist that uses php arrays. JSON would be even more portable, but possibly more limiting too (no built-in static checks).

https://gist.github.com/ewinslow/6445c79c9d0db60ed2dd
## Why?

There's a lot of talk about changing up the schema lately and I just feel like it would be nice to make this process easier on ourselves...
### Track diffs with git, not PHP classes

Several other frameworks support DB migrations, but they take the manual up/down diffing approach. This means there is nowhere in the code where you can look at the current schema. You must deduce it from a sequence of migrations. It is like having a form of version control on top of the one git is already giving us. Wouldn't it be better to just view the diffs with git, and not write them manually? Besides that, `down` strikes me as horribly dangerous and prone to data loss. Migrations should always be safe by default. A framework that performs diffs for you can do this automatically.
### Write the schema you want, let the framework figure out how to get there

The current schema upgrade process in Elgg involves writing both the final schema _and_ the diff from the previous schema. This seems horribly error prone since you can write an upgrade script differently from the final product. It also puts me off doing any schema migrations since I essentially have to put twice as much effort into making sure the migration is bug-free.
### Declarative foreign keys

MySQL requires you to write foreign keys as add/drop commands in a certain order with the tables that are present. This is less than ideal as the foreign key definitions then live farther away from the place where they are relevant (i.e. the table on which they apply). It would be easier (IMO), to write foreign keys in a more declarative fashion and let the framework figure out in what order it should add them.
### Better defaults
- charset = utf8
- collation = utf8_general_ci
- engine = innodb
- nullable = false

If we rely on raw SQL, we cannot rely on any defaults since those may be different depending on the user's server configuration and MySQL version. This also means we can default to InnoDB/UTF8 and not have that information cluttering the schema source code.

### LINKED ISSUE #10896: River table schema missing subtype key

### PR DESCRIPTION
**Pull request by hypeJunction**
**feat(db): database migrations are now managed by phinx**

Elgg now uses phinx for schema migrations.

Migrations are now run before code is booted in upgrade.php

Fixes #7947
Fixes #5442

**fix(db): add missing subtype index to river table**

Fixes #10896

### CODE REVIEW

**Comment by jdalsem** [comment]
Looks nice! This will replace the old migration files?

**Comment by hypeJunction** [comment]
No, this is a footprint of the current database state.
I suppose we could roll back to pre-master state, and add new migrations using this.

**Comment by hypeJunction** [comment]
Should we let plugins register their migrations from get go or wait until we have a compelling use case?

**Comment by jdalsem** [comment]
i never had a plugin that needed database manipulation so lets wait. I do not want to encourage custom db stuff

**Comment by hypeJunction** [comment]
I have plugins that add their own tables, so it would be handy to just allow them to use core interface, but we can do that later.

I have added some basic docs.

**Comment by jdalsem** [comment]
what happens if you perform an upgrade... will the migration files be executed? How does Phinx know which migrations need to be applied?

**Comment by hypeJunction** [comment]
> what happens if you perform an upgrade... will the migration files be executed? 

Migrations are executed on each upgrade and on install.

> How does Phinx know which migrations need to be applied?

It creates a new database table and stores migration names and timestamps

**Comment by jdalsem** [comment]
I believe this fixes https://github.com/Elgg/Elgg/issues/5442

**Comment by hypeJunction** [comment]
Need to check if it is already so, but definitely can definitely run migrations without booting core

**Comment by hypeJunction** [comment]
So, this runs before upgrade.php now. Added a new migration for the river table, so you can see how easy it is to make changes.

We should probably wipe mysql.sql, or do we want to update it for each migration manually?

File: engine/schema/migrations/20170802080931_add_subtype_index_to_river_table.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	 *    addForeignKey
+	 *
+	 * Remember to call "create()" or "update()" and NOT "save()" when working
+	 * with the Table class.
+	 */
+	public function change() {
+
+		$table = $this->table('river');
+
+		$table->addIndex(['type'], [
```
Comment (hypeJunction) [inline engine/schema/migrations/20170802080931_add_subtype_index_to_river_table.php]: should be `subtype` though

**Comment by hypeJunction** [comment]
Should we pull the trigger on this? I don't want to start any work on subtypes before this is in.

**Comment by hypeJunction** [comment]
Let me try and roll back to 2.3 start and see if we can do all migrations with phinx for 3.0

**Comment by hypeJunction** [comment]
There were too many schema changes. Let's leave at current state.

**Comment by jdalsem** [comment]
i am going to do some testing.. if that works out ok, we can merge it

File: engine/classes/Elgg/Application.php [hunk 2]
```diff
@@ -569,6 +572,62 @@ public static function upgrade() {
...
+		return true;
+	}
+
+	/**
+	 * Returns configuration array for database migrations
+	 * @return array
+	 */
+	public static function getMigrationSettings() {
+		$config = self::create()->services->config;
+		$config->loadSettingsFile();
```
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: Note to self: I'll need to rewrite this function a bit for #11092. Something like:

```php
$config = self::factory()->services->config;
...
$config->dbprefix; // getVolatile going away
```

**Comment by jdalsem** [comment]
upgrading fails from a fresh 2.3 install... maybe unrelated to this PR ... will investigate a bit more

**Comment by jdalsem** [comment]
Upgrade issue here #11103

**Comment by jdalsem** [comment]
upgrade issue fixed . rebase and try again

**Comment by mrclay** [comment]
FYI I'm trying to rebase/alter this as needed.

**Comment by mrclay** [comment]
Can #11121 replace this?

**Comment by mrclay** [comment]
Done in #11121

**Pull request closed without merging**